### PR TITLE
CLDC-2862: Prepare maintenance page for PaaS app

### PR DESCRIPTION
**Pull request by SamSeed-Softwire**
https://dluhcdigital.atlassian.net/browse/CLDC-2862

This implementation may be a little naive, but given how experienced people on the team are with Rails I thought it made sense to raise a PR sooner rather than later.

### CODE REVIEW

File: app/controllers/application_controller.rb
Comment (SamSeed-Softwire) [inline app/controllers/application_controller.rb]: The addition of the `request.fullpath.split("?")[0].delete("/") != "service-unavailable"` condition was to stop the app getting stuck in a loop (it wouldn't load the page without this).

File: app/services/feature_toggle.rb
Comment (SamSeed-Softwire) [inline app/services/feature_toggle.rb]: I guess we might want to use an environment variable here instead? That can be set to true or false in the deployment pipelines. Wanted to sanity check the best approach with others before implementing that though.

File: app/views/maintenance/service_unavailable.html.erb
Comment (SamSeed-Softwire) [inline app/views/maintenance/service_unavailable.html.erb]: I've asked Susanna to confirm she's happy with this content.

File: app/views/maintenance/service_unavailable.html.erb
Comment (SamSeed-Softwire) [inline app/views/maintenance/service_unavailable.html.erb]: Here's my current understanding of whether answers get saved or not:
- If maintenance mode gets activated whilst a user is on a page, anything they've done on that page (e.g. making a change to an answer on a log, or submitting a bulk upload) will be lost.
- Otherwise data is saved.

That seems to be the case from my testing and my understanding of what's going on, but I wanted to check here with people more experienced with Rails and this codebase in case there are any edge cases I'm missing.

Assuming I'm right, what do people think would be sensible content here? "We have not saved any changes you made immediately prior to the site undergoing maintenance"?

File: app/views/maintenance/service_unavailable.html.erb
Comment (SamSeed-Softwire) [inline app/views/maintenance/service_unavailable.html.erb]: This will need confirming and updating.

File: spec/requests/maintenance_controller_spec.rb
Comment (SamSeed-Softwire) [inline spec/requests/maintenance_controller_spec.rb:75]: I wasn't able to simulate it redirecting 'back' (in `app/controllers/application_controller.rb` we have `redirect_back(fallback_location: root_path)`). I think this is fine - I'm not trying to test that the `redirect_back` method works - I'm trying to test that the redirect gets triggered correctly.

File: spec/requests/maintenance_controller_spec.rb
Comment (SamSeed-Softwire) [inline spec/requests/maintenance_controller_spec.rb:24]: I haven't tested that the user is signed out here, because I tested that in `spec/controllers/maintenance_controller_spec.rb`.

File: app/views/maintenance/service_unavailable.html.erb
Comment (RachaelBooth) [reply app/views/maintenance/service_unavailable.html.erb]: I agree with your understanding on what does and doesn't get saved. 

I think if I was in the middle of filling out a log though and was told that my answers hadn't been saved (including with any immediately prior wording), I'd expect the entire log's worth of answers not to be saved. Although I guess the buttons on each question do say "save and continue". Probably worth asking Susannah about wording.

**Review by RachaelBooth — COMMENTED** [review]
Were we going to disable scheduled jobs as part of this, or is that for a future PR?

**Comment by RachaelBooth** [comment]
Presumably testing this will require some co-ordination of people using the review app and turning the maintenance mode on.

**Comment by SamSeed-Softwire** [comment]
> Presumably testing this will require some co-ordination of people using the review app and turning the maintenance mode on.

I tested it locally by doing this and was happy with everything. But yes I imagine whoever tests this will want to do the same.

**Comment by SamSeed-Softwire** [comment]
> Were we going to disable scheduled jobs as part of this, or is that for a future PR?

@RachaelBooth ah, forgot about that (it's not mentioned in the ticket and I don't think it's in any other tickets btw). I will implement now as part of this PR.

**Comment by RachaelBooth** [comment]
> > Were we going to disable scheduled jobs as part of this, or is that for a future PR?
> 
> @RachaelBooth ah, forgot about that (it's not mentioned in the ticket and I don't think it's in any other tickets btw). I will implement now as part of this PR.

It's vaguely mentioned in a comment on this ticket as an "either here or another ticket" thing - but we've not got another ticket for it.

File: app/controllers/application_controller.rb
Comment (SamSeed-Softwire) [inline app/controllers/application_controller.rb:10]: The exclusion of the `service-unavailable` path was to stop the app getting stuck in a loop (it wouldn't load the page without this).

Comment (SamSeed-Softwire) [inline app/controllers/application_controller.rb:10]: I've also allowed the user to interact with the cookie banner and see the cookies page, as well as the accessibility statement and privacy notice, as I assumed these are bits of information that we should probably make available even in maintenance mode. If you disagree and would prefer these to be hidden I can also do that.

File: config/initializers/sidekiq.rb
Comment (SamSeed-Softwire) [inline config/initializers/sidekiq.rb]: Without this line, jobs will not be destroyed during a deployment. So enabling maintenance mode wouldn't turn off the data export xml job. Also on staging right now there is still the data export csv job running (and failing) every day - it would be good to clean bits like that up.

We could probably also destroy the jobs in the startup block, but I don't think it matters. Also we could specify specific jobs to destroy instead of all of them, but it seems wise to me to destroy everything!

File: app/controllers/application_controller.rb
Comment (SamSeed-Softwire) [reply app/controllers/application_controller.rb:10]: Note I haven't written tests for this because I wanted to get confirmation first that we did indeed want to retain these links in maintenance mode.

File: config/initializers/sidekiq.rb
Comment (SamSeed-Softwire) [reply config/initializers/sidekiq.rb]: Note that AFACIT there were no tests for this sidekiq config/job stuff. Seems wise for me to write some, shout if you think that's not worth the time right now though.

Comment (RachaelBooth) [reply config/initializers/sidekiq.rb]: Does this apply every time an app instance is shut down?

File: app/controllers/application_controller.rb
Comment (RachaelBooth) [reply app/controllers/application_controller.rb:10]: I think that makes sense.

File: config/initializers/sidekiq.rb
Comment (SamSeed-Softwire) [reply config/initializers/sidekiq.rb]: As agreed, we want to move this line into the server startup section. There's a chance that if a server starts the moment before the job is supposed to run then the job will get deleted and not recreated in time so the job won't get run. We think this is a tiny risk and is easily outweighed by the advantages of ensuring the jobs which exists are aligned with what's defined in the code.

File: app/services/feature_toggle.rb
Comment (SamSeed-Softwire) [reply app/services/feature_toggle.rb]: When we enable it, we'll do it by specifying which environments it should be enabled in (like all the other feature toggles).

File: app/controllers/application_controller.rb
Comment (SamSeed-Softwire) [reply app/controllers/application_controller.rb:10]: I've written tests for these now.

File: spec/requests/maintenance_controller_spec.rb
Comment (SamSeed-Softwire) [inline spec/requests/maintenance_controller_spec.rb:80]: I had to follow the redirect twice here in order to reach a page which wasn't just a "You are being redirected." page

File: spec/requests/maintenance_controller_spec.rb
Comment (SamSeed-Softwire) [inline spec/requests/maintenance_controller_spec.rb:27]: I could imagine it being sensible to move these cookie banner tests into another file (e.g. `spec/views/layouts/application_layout_spec.rb`, although I'm not convinced that is the right place) but I thought they were probably fine here. Happy to be persuaded otherwise though.

File: spec/requests/cookies_controller_spec.rb
Comment (SamSeed-Softwire) [inline spec/requests/cookies_controller_spec.rb:3]: There are other aspects of the cookies controller which I'm not testing here (some of the actual behaviour of the banner), but I thought that was out of scope for this ticket.

File: config/initializers/sidekiq.rb
Comment (SamSeed-Softwire) [inline config/initializers/sidekiq.rb:33]: There were no existing tests for this initializer file (nor for most initialisers). I'm not convinced writing a test here to check that Sidekiq can start up a job successfully is that worth the time. There's obviously the question of whether the DataExportXMLJob class works correctly (and there's already a test for that) so I guess the final piece would be to check that `config/sidekiq_cron_schedule.yml` is defined correctly, however I don't think that's worth the time investment now (if we do want to do that, I think we should raise a separate ticket for that). Do you agree?

Comment (RachaelBooth) [reply config/initializers/sidekiq.rb:33]: That's definitely out of scope for this (given there are currently no tests and it's not really a thing where there's an obviously worthwhile test to do) - it's an app team problem to deal with later (or to decide not to).

Comment (SamSeed-Softwire) [reply config/initializers/sidekiq.rb:33]: I raised a ticket for this then: https://dluhcdigital.atlassian.net/browse/CLDC-2994.

In it I do mention that we might not want tests at all - it's more a placeholder for making the decision, so it doesn't get forgotten.

**Review by RachaelBooth — APPROVED** [review]

**Comment by RachaelBooth** [comment]
Should make sure someone explicitly does some testing of this before merging (I've looked at the code, but not the review app with the maintenance mode on)

**Comment by SamSeed-Softwire** [comment]
> Should make sure someone explicitly does some testing of this before merging (I've looked at the code, but not the review app with the maintenance mode on)

@RachaelBooth I've tested this and happy with it, but it should probably have another pair of eyes on it - will ask the team.

**Comment by SamSeed-Softwire** [comment]
> > Should make sure someone explicitly does some testing of this before merging (I've looked at the code, but not the review app with the maintenance mode on)
> 
> @RachaelBooth I've tested this and happy with it, but it should probably have another pair of eyes on it - will ask the team.

@RachaelBooth Nat has tested this and added his comments to the [ticket](https://dluhcdigital.atlassian.net/browse/CLDC-2862). As agreed he has not tested the sidekiq job enqueueing functionality.

**Pull request merged**
